### ISSUE TITLE
Add Microseconds and Nanoseconds to TimeStamp, DateTime, and DateTimeOffset

### ISSUE THREAD

**Issue by ChristopherHaws** [dotnet/corefx#24555]
Currently, the value of a single `Tick` in .NET is 100ns, however there is no API surface for accessing microseconds or nanoseconds other than using `Ticks` to manually calculate these numbers. Seems like this could be done in a minor version since it would be adding API surface, not changing existing API's, and functionality would be unchanged.

# Proposed API
```diff
namespace System {
     public struct DateTime {
+        public DateTime(int year, int month, int day, int hour, int minute, int second, int millisecond, int microseconds);
+        public DateTime(int year, int month, int day, int hour, int minute, int second, int millisecond, int microseconds, System.DateTimeKind kind);
+        public DateTime(int year, int month, int day, int hour, int minute, int second, int millisecond, int microseconds, System.Globalization.Calendar calendar);
+        public DateTime(int year, int month, int day, int hour, int minute, int second, int millisecond, int microseconds, int nanoseconds);
+        public DateTime(int year, int month, int day, int hour, int minute, int second, int millisecond, int microseconds, int nanoseconds, System.DateTimeKind kind);
+        public DateTime(int year, int month, int day, int hour, int minute, int second, int millisecond, int microseconds, int nanoseconds, System.Globalization.Calendar calendar);
+        public int Microsecond { get; }
+        public int Nanosecond { get; }
+        public System.DateTime AddMicroseconds(double microseconds);
+        public System.DateTime AddNanoseconds(double nanoseconds);
     }
     public struct DateTimeOffset {
+        public DateTimeOffset(int year, int month, int day, int hour, int minute, int second, int millisecond, int microseconds, System.TimeSpan offset);
+        public DateTimeOffset(int year, int month, int day, int hour, int minute, int second, int millisecond, int microseconds, System.TimeSpan offset, System.Globalization.Calendar calendar);
+        public DateTimeOffset(int year, int month, int day, int hour, int minute, int second, int millisecond, int microseconds, int nanoseconds, System.TimeSpan offset);
+        public DateTimeOffset(int year, int month, int day, int hour, int minute, int second, int millisecond, int microseconds, int nanoseconds, System.TimeSpan offset, System.Globalization.Calendar calendar);
+        public int Microsecond { get; }
+        public int Nanosecond { get; }
+        public System.DateTimeOffset AddMicroseconds(double microseconds);
+        public System.DateTimeOffset AddNanoseconds(double nanoseconds);
+        public static System.DateTimeOffset FromUnixTimeMicroseconds(long microseconds);
+        public static System.DateTimeOffset FromUnixTimeNanoseconds(long nanoseconds);
+        public static long ToUnixTimeMicroseconds();
+        public static long ToUnixTimeNanoseconds();
     }
     public struct TimeSpan {
+        public const long TicksPerMicrosecond = 10L;
+        public const double TicksPerNanosecond = 0.01D;
+        public TimeSpan(int days, int hours, int minutes, int seconds, int milliseconds, int microseconds);
+        public TimeSpan(int days, int hours, int minutes, int seconds, int milliseconds, int microseconds, int nanoseconds);
+        public double Microseconds { get; }
+        public double Nanoseconds { get; }
+        public double TotalMicroseconds { get; }
+        public double TotalNanoseconds { get; }
+        public static System.TimeSpan FromMicroseconds(double value);
+        public static System.TimeSpan FromNanoseconds(double value);
     }
 }
```

# Example Usage
```csharp
public class Samples
{
    public void DateTimeSamples()
    {
        new DateTime(0001, 01, 01, 00, 00, 00, 00, 999).Ticks;      // 9990
        new DateTime(0001, 01, 01, 00, 00, 00, 00, 00, 900).Ticks;  // 9
        DateTime.Parse("0001-01-01 00:00:00.0009990").Microsecond;  // 999
        DateTime.Parse("0001-01-01 00:00:00.0000009").Nanosecond;   // 900
        DateTime.Zero.AddMicroseconds(999).Ticks;                   // 9990
        DateTime.Zero.AddNanoseconds(900).Ticks;                    // 900
    }

    public void DateTimeOffsetSamples()
    {
        new DateTimeOffset(0001, 01, 01, 00, 00, 00, 00, 999, TimeSpan.FromHours(-7)).Ticks;        // 9990
        new DateTimeOffset(0001, 01, 01, 00, 00, 00, 00, 00, 900, TimeSpan.FromHours(-7)).Ticks;    // 9
        DateTimeOffset.Parse("0001-01-01 00:00:00.0009990 -7").Microsecond;                         // 999
        DateTimeOffset.Parse("0001-01-01 00:00:00.0000009 -7").Nanosecond;                          // 900
        new DateTimeOffset().AddMicroseconds(999).Ticks;                                            // 9990
        new DateTimeOffset().AddNanoseconds(900).Ticks;                                             // 900
        DateTimeOffset.FromUnixTimeMicroseconds(999).Ticks;                                         // 9990
        DateTimeOffset.FromUnixTimeNanoseconds(900).Ticks;                                          // 9
        DateTimeOffset.Parse("1970-01-01 00:00:00.0009990Z").ToUnixTimeMicroseconds();              // 999
        DateTimeOffset.Parse("1970-01-01 00:00:00.0000009Z").ToUnixTimeNanoseconds();               // 900
    }

    public void TimeSpanSamples()
    {
        new TimeSpan(999 * TimeSpan.TicksPerMicrosecond).Ticks;	        // 9990
        new TimeSpan((long)(900 * TimeSpan.TicksPerNanosecond)).Ticks;  // 9
        new TimeSpan(0001, 01, 01, 00, 00, 00, 00, 999).Ticks;          // 9990
        new TimeSpan(0001, 01, 01, 00, 00, 00, 00, 00, 900).Ticks;      // 9
        TimeSpan.Parse("0001-01-01 00:00:00.0009990").Microsecond;      // 999
        TimeSpan.Parse("0001-01-01 00:00:00.0000009").Nanosecond;       // 900
        TimeSpan.Zero.AddMicroseconds(999).Ticks;                       // 9990
        TimeSpan.Zero.AddNanoseconds(900).Ticks;                        // 900
    }
}
```

**Comment by svick** [comment]
Is the common use case really to have separate values for seconds, milliseconds, microseconds and nanoseconds? Wouldn't it be better to instead expose a floating point number that's the number of seconds, including the fractional part?

I realize this would be inconsistent (seconds and milliseconds are already separate), but I think usefulness beats consistency.

Though some of the proposed members (like `AddMicroseconds` or `TotalMicroseconds`) would still make sense.

(Also, I'm not sure what exactly would be the right API to expose that.)

**Comment by tarekgh** [comment]
> Is the common use case really to have separate values for seconds, milliseconds, microseconds and nanoseconds? Wouldn't it be better to instead expose a floating point number that's the number of seconds, including the fractional part?

If the caller has the values of seconds, milliseconds, microseconds, nanoseconds, would need to do some calculations to convert these values into the second fraction to be able to call the API. I think the request here is to help users not do the calculations themselves. users can achieve the result to do by calculating the ticks from the second fractions. so I am seeing the proposed APIs just a helper methods.

**Comment by tarekgh** [comment]
@ChristopherHaws what is the need for the To/From Unix APIs? could you please clarify how these exactly get used?

**Comment by ChristopherHaws** [comment]
@tarekgh Correct, there is no functional changes. It is just additional properties and methods for making it easier to work with microseconds and nanoseconds, which already exists in the `Ticks`. Here is a recent PR where I made some helper methods for doing just this: https://github.com/fluentassertions/fluentassertions/pull/669/files

Specificly `TimeSpanConversionExtensions.cs` and `FluentDateTimeExtensions.cs`

**Comment by svick** [comment]
@tarekgh 

> If the caller has the values of seconds, milliseconds, microseconds, nanoseconds, would need to do some calculations to convert these values into the second fraction to be able to call the API.

I get that. My question is: how often does that happen? As an example, what's more likely, that the input is "3 minutes 14 seconds 159 milliseconds" or "3 minutes 14.159 seconds"? I think it's the latter and the current API does not handle that case well and the API proposed here does not improve it much either.

**Comment by ChristopherHaws** [comment]
@svick I think that both ways are valid for different reasons. The way I proposed in the issue are simply extending the existing API's standard. If `DateTime.Seconds` had been a method instead, I would say there should be an overload that allows you to decide if it includes the higher precision values or not. I cant think of a good name, but something like might work better for what you are talking about:

```csharp
namespace System {
    public struct DateTime {
        public double GetPart(DateTimePart.Seconds, DateTimePartOptions.IncludeHigherPrecision);
    }
}
```

**Comment by tarekgh** [comment]
@ChristopherHaws 

- I am not seeing a big value to have From/To Unix proposed APIs. if you agree, we can remove them from the proposal.
- Also in DateTime/TimeSpan, the minimum unit we are using is the ticks (which is long value and not double). 1 Tick is 100 nanoseconds. so the proposed APIs that takes nanoseconds is kind doesn't make sense except if we ask users passing it as multiple of 100 nanoseconds. and that can complicate the APIs and make it inconsistent with the rest. Users can easily add ticks which covers adding 100 nanoseconds. I suggest removing all proposed APIs which is dealing with the nanoseconds from the proposal too. 

please let me know what you think.

**Comment by Clockwork-Muse** [comment]
> I am not seeing a big value to have From/To Unix proposed APIs. if you agree, we can remove them from the proposal.

Um, but what about #24449 ?

**Comment by tarekgh** [comment]
@Clockwork-Muse for #24449, I am seeing a lot of people is using 

```C#
new DateTime(1970, 1, 1, 0, 0, 0, DateTimeKind.Utc)
```

this why #24449 would have a value. but for the proposed Unix APIs here, I am not seeing many people using it.

**Comment by ChristopherHaws** [comment]
@tarekgh 
- I am ok with removing the From/To Unix API's. At some point I could see the value in having To/From UnixMicroseconds, but I would say that they are much less important than the rest of the helpers, so I will remove them from the proposal.
- I can see the issues with adding nanoseconds (mainly needing to add in 100ns increments). I wish there were a better name that could be given to the method to make that more obvious, but I cant really think of one. Would a valid compromise be to keep the read apis, but remove the write apis? For example, keep `Nanoseconds { get; }` but remove `FromNanoseconds(double value);` and the constructors containing nanoseconds. I would be ok with this, but I still don't like that as a consumer of `DateTime`, if I want to add nanoseconds, I need to know how to convert between ticks and nanoseconds. For this reason, I would recommend we keep the `TicksPerNanosecond` const.
- I fixed the doubles, thanks for pointing this out.
  - `DateTime.Microsecond` and `DateTime.Nanosecond` return int now since the values must be between 0 and 999.
  - `DateTime.AddMicroseconds` takes an int now since the value must be between 0 and 999.
  - `DateTimeOffset.Microsecond` and `DateTimeOffset.Nanosecond` return int now since the values must be between 0 and 999.
  - `DateTimeOffset.AddMicroseconds` takes an int now since the value must be between 0 and 999.
  - `TimeSpan.Microseconds` and `TimeSpan.Nanoseconds` return int now since the values must be between 0 and 999.
  - `TimeSpan.TotalMicroseconds` and `TimeSpan.TotalNanoseconds` return double still since that is what the other total properties return. This could be long but I kept it as double for consistency. Let me know if I should change this.
  - `TimeSpan.FromMicroseconds` takes a double still to keep the api consistent with the other `From` methods.

**Removed from Proposal**
```csharp
namespace System {
    public struct DateTime {
        public DateTime(int year, int month, int day, int hour, int minute, int second, int millisecond, int microseconds, int nanoseconds);
        public DateTime(int year, int month, int day, int hour, int minute, int second, int millisecond, int microseconds, int nanoseconds, System.DateTimeKind kind);
        public DateTime(int year, int month, int day, int hour, int minute, int second, int millisecond, int microseconds, int nanoseconds, System.Globalization.Calendar calendar);
        public System.DateTime AddNanoseconds(int nanoseconds);
    }
    public struct DateTimeOffset {
        public DateTimeOffset(int year, int month, int day, int hour, int minute, int second, int millisecond, int microseconds, int nanoseconds, System.TimeSpan offset);
        public DateTimeOffset(int year, int month, int day, int hour, int minute, int second, int millisecond, int microseconds, int nanoseconds, System.TimeSpan offset, System.Globalization.Calendar calendar);
        public static System.DateTimeOffset FromUnixTimeMicroseconds(long microseconds);
        public static System.DateTimeOffset FromUnixTimeNanoseconds(long nanoseconds);
        public System.DateTimeOffset AddNanoseconds(int nanoseconds);
        public static long ToUnixTimeMicroseconds();
        public static long ToUnixTimeNanoseconds();
    }
    public struct TimeSpan {
        public static System.TimeSpan FromNanoseconds(double value);
        public TimeSpan(int days, int hours, int minutes, int seconds, int milliseconds, int microseconds, int nanoseconds);
    }
 }
```

**Comment by tarekgh** [comment]
@ChristopherHaws thanks for your updates. I'll try to look more get back to you if I have any other comment. could you please move the allowed values of the millisecond, Microseconds...etc. to the proposal, section to be clear about how the APIs will work and when we throw?

**Comment by ChristopherHaws** [comment]
@tarekgh Added in the form of xml docs to the proposal. Thanks!

**Comment by tarekgh** [comment]
@iamcarbon I am seeing your reaction to the proposal. do you have any specific feedback to understand why you don't like the proposal? Thanks.

**Comment by iamcarbon** [comment]
@tarekgh DateTime is only accurate to 100 nanoseconds. It's confusing to have a property that cannot represent its full range.

**Comment by tarekgh** [comment]
> It's confusing to have a property that cannot represent its full range.

I am not sure I fully understand this but DateTime has Ticks property that can reflect the ticks (in hundreds of nanoseconds). Also, it has MinValue and MaxValue properties which can tell the min and max of the supported range. if you are referring to a specific property in the proposal, please let me know so we can evaluate your feedback.

**Comment by ChristopherHaws** [comment]
@tarekgh I believe he is referring to the fact that the `Nanoseconds` property would only ever return one of 10 values (0, 100, 200, 300, 400, 500, 600, 700, 800, 900).

I personally don't have any issue with this, as the alternative is that anyone who needs to know the amount of nanoseconds would have to calculate it from the ticks, but it should definitely be pointed out in the documentation. IMO, since `DateTime` does have nanoseconds in it's precision, a consumer should be able to access it as a property.

**Comment by tarekgh** [comment]
This is a limitation of DateTime. even if we were supporting nanoseconds we could get cases asking for femtoseconds :-) 

At any case, we have to have a minimum unit we can support considering we want to keep DateTime size not exceeding long size for perf reasons.

**Comment by ChristopherHaws** [comment]
@tarekgh I agree. I think he was just pointing out that the signature states that it returns an int between 0 and 900, but the values are not contiguous which might confuse some people that don't realize that DateTime's precision is 100ns. I think asking users to convert ticks manually is much more confusing and potentially error prone.

**Comment by danmosemsft** [comment]
@ChristopherHaws of course a hypothetical future runtime/platform could support a finer resolution with the same API, and I don't believe we would consider that a breaking change.

**Comment by ChristopherHaws** [comment]
@danmosemsft All the more reason for these helper methods/properties. If someone is making an assumption about the resolution of ticks, their code would break if that resolution changed. With these helper methods, consumers can be sure that `TimeSpan.Microseconds` returns microseconds and not nanoseconds or picoseconds. :)

**Comment by tarekgh** [comment]
@joperezr this proposal looks good to me. if you agree, please mark the issue as ready for design review. Thanks.

**Comment by mj1856** [comment]
Be *very* careful with this.  Exposing micros and nanos directly can easily lead to overflow exceptions or floating point errors.

For example, `TimeSpan.MaxValue.TotalNanoseconds` would give a `double` of `9.22337203685477E+20`, which cannot be converted to a `long` without overflowing.  Essentially this is the same as:

```csharp
(long) (TimeSpan.MaxValue.TotalSeconds * 1000 * 1000 * 1000)  // -9223372036854775808
```

There's also a loss of precision with such extreme floating point values - which is counter-intuitive when working with APIs that promise nanoseconds.

I advise caution.  We don't want people to think `DateTime`, `DateTimeOffset` and `TimeSpan` now have increased precision.

**Comment by ChristopherHaws** [comment]
@mj1856 Agreed. I was thinking about this as well. From what I can see, TotalMicroseconds shouldn't have any issue, but TotalNanoseconds could have issues when you start approaching and `TimeSpan.Max` and `TimeSpan.Min`. Please correct me if I am wrong about this. I will add an overflow exception to the original post so that it is not overlooked. If you have any thoughts on a better way to handle this, I am all ears! :)

**Comment by tarekgh** [comment]
what about making TotalNanoseconds returns double instead?

**Comment by ChristopherHaws** [comment]
@tarekgh Do you mean `long`? The only reason it returns `double` in the current proposal is because the other `Total` properties return `double`'s.

**Comment by tarekgh** [comment]
@ChristopherHaws ignore my comment. I was looking at different place :-(

**Comment by ChristopherHaws** [comment]
@mj1856 I was wrong, there is a potential for loss of precision with `double` even with microseconds. One alternative could be to use `decimal` since it's 128-bit. Thoughts?

```csharp
void Main()
{
    TimeSpan.MaxValue.TotalMicroseconds();          // 9.22337203685478E+17
    TimeSpan.MaxValue.TotalMicrosecondsAsDecimal(); // 922337203685477580.7
    TimeSpan.MaxValue.TotalNanoseconds();           // 9.22337203685478E+20
    TimeSpan.MaxValue.TotalNanosecondsAsDecimal();  // 922337203685477580700
}

public static class TimeSpanEx
{
    public const Int64 TicksPerMicrosecond = TimeSpan.TicksPerMillisecond / 1000;
    public const Double TicksPerNanosecond = TicksPerMicrosecond / 1000d;

    public static Double TotalMicroseconds(this TimeSpan self) => self.Ticks * (1d / TicksPerMicrosecond);
    public static Decimal TotalMicrosecondsAsDecimal(this TimeSpan self) => self.Ticks * (1m / TicksPerMicrosecond);

    public static Double TotalNanoseconds(this TimeSpan self) => self.Ticks * (1d / TicksPerNanosecond);
    public static Decimal TotalNanosecondsAsDecimal(this TimeSpan self) => self.Ticks * (1m / (Decimal)TicksPerNanosecond);
}
```

**Comment by mj1856** [comment]
Well, the fact is that all three types in question already have their range locked in, and over that range one cannot represent nanoseconds in 64 bits without losing some of them.  Floating point type allow for loss of precision, but one needs `System.Numerics.BigInteger` to truly work with nanoseconds without any loss. 
 We're also touching on why ticks go to 7 decimals and no further.  8 decimals and you start losing information over this range.

I'm not saying don't do this, I'm just saying check that the APIs can handle everything at all boundaries, especially with nanoseconds.

**Comment by joperezr** [comment]
Given the above concerns about overflow and precision, let's discuss a bit more on what would the right return type should be.

**Comment by tarekgh** [comment]
My take here: 

-  for using Decimal type, I think double is covering the whole range. I am not sure if we are losing any precision here.  @ChristopherHaws, when you tried to print the calculated result when using double, did you use G17 formatting? 
- We should have the documentation to warn the consumer if they want to use integral types (e.g long), this can cause overflow.

**Comment by ChristopherHaws** [comment]
@tarekgh Here are the results using G17 formatting:

```csharp
void Main()
{
    TimeSpan.MaxValue.TotalMicroseconds().ToString("G17", CultureInfo.InvariantCulture);            // 9.2233720368547763E+17
    TimeSpan.MaxValue.TotalMicrosecondsAsDecimal().ToString("G17", CultureInfo.InvariantCulture);   // 9.2233720368547758E+17
    TimeSpan.MaxValue.TotalNanoseconds().ToString("G17", CultureInfo.InvariantCulture);             // 9.2233720368547758E+20
    TimeSpan.MaxValue.TotalNanosecondsAsDecimal().ToString("G17", CultureInfo.InvariantCulture);    // 9.2233720368547758E+20
}

public static class TimeSpanEx
{
    public const Int64 TicksPerMicrosecond = TimeSpan.TicksPerMillisecond / 1000;
    public const Double TicksPerNanosecond = TicksPerMicrosecond / 1000d;

    public static Double TotalMicroseconds(this TimeSpan self) => self.Ticks * (1d / TicksPerMicrosecond);
    public static Decimal TotalMicrosecondsAsDecimal(this TimeSpan self) => self.Ticks * (1m / TicksPerMicrosecond);

    public static Double TotalNanoseconds(this TimeSpan self) => self.Ticks * (1d / TicksPerNanosecond);
    public static Decimal TotalNanosecondsAsDecimal(this TimeSpan self) => self.Ticks * (1m / (Decimal)TicksPerNanosecond);
}
```

**Comment by tarekgh** [comment]
I believe this is accepted precision.

**Comment by ChristopherHaws** [comment]
Is there any particular reason not to use `decimal`? If we have the ability to have no lose of precision using a built in type, I think we should consider using it.

**Comment by tarekgh** [comment]
I am not seeing the precision here is the issue. the more concern I have is what @mj1856 raised which is the possible overflow. 

for example, imagine someone wants to get the total microseconds of TimeSpan.MaxValue which will produce ```9.2233720368547763E+17``` (using double). then want to round trip this value back to ticks. 

doing 

```C#
(long) (TimeSpan.MaxValue.TotalMicroseconds() * TimeSpanEx.TicksPerMicrosecond)
```

will result in overflow. and if did 

```C#
(ulong) (TimeSpan.MaxValue.TotalMicroseconds() * TimeSpanEx.TicksPerMicrosecond)
```
will produce wrong ticks which would be a negative number if you convert it to long.

**Comment by Thaina** [comment]
@tarekgh They was in needed in firestore. Please stop forcing people to use that ugly `new DateTime(1970, 1, 1, 0, 0, 0, DateTimeKind.Utc)` anywhere

https://github.com/googleapis/google-cloud-dotnet/issues/2546#issuecomment-450796398

**Comment by tarekgh** [comment]
Thanks @Thaina for pointing at that. reading the comments there, it looks they have some limitation to run on versions which doesn't support FromUnixTimeSeconds. I think this is a matter of time to get into the point always running on versions support FromUnixTimeSeconds and then we should force using it. please keep pointing to such cases and we can help pushing if there is no good reason not to use it.

**Comment by Thaina** [comment]
@tarekgh Versioning is one problem but the main point of discussion there is, firestore system use a number for time in unix time value but more precise than millisecond. Even they would use latest version they still couldn't use `FromUnixTimeMilliSeconds`

I just pointed out that there are really a system that keep unix timestamp in very precise format. It also somewhat popular system in very large and firm like google. So `FromUnixTimeMicroSeconds` `FromUnixTimeNanoSeconds` and maybe `FromUnixTimeTicks` should be available

**Comment by xifedi7** [comment]
Hi I'm a new to this. Working with an API where I get now data but with a for my unclear time stamp (format) e.g.: "636926976000000000" & "636927840000000000" 
The time difference is 24h ,I verified that.
How can I this converted to a date/time that make any sense ?

**Comment by tarekgh** [comment]
do the following:

```C#
            Console.WriteLine(new DateTime(636926976000000000)); // or new DateTime(636926976000000000).ToString() if you want to get it as string
            Console.WriteLine(new DateTime(636927840000000000)); 
```

This will print it as:

```
5/6/2019 12:00:00 AM
5/7/2019 12:00:00 AM
```
